File: muxdb-py/muxdb/router.py
"""
Query router — Layer 1 core.

Parses incoming queries, extracts shard keys, resolves the target shard(s)
via the ShardMap, and orchestrates scatter-gather for cross-shard queries.

Implements the Vitess VTGate / Slicer Clerk routing model: the application
sees a single logical database, the router handles all physical topology.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any, Sequence

import structlog

from muxdb.config import MuxConfig
from muxdb.errors import RoutingError, CrossShardTransactionError
from muxdb.shard_map import ShardMap, ShardInfo, create_shard_map

logger = structlog.get_logger(__name__)


class RouteType(Enum):
    """Classification of how a query should be executed."""

    SINGLE = auto()       # Routed to exactly one shard
    SCATTER = auto()      # Sent to all shards, results gathered
    BROADCAST = auto()    # Sent to all shards (DDL, no result merge)


@dataclass(frozen=True)
class RouteDecision:
    """The outcome of routing a query."""

    type: RouteType
    targets: list[ShardInfo]
    shard_key_value: str | int | None = None
    query: str = ""
    params: tuple[Any, ...] = ()
    elapsed_us: float = 0.0  # microseconds spent routing


@dataclass
class QueryContext:
    """Metadata about the current query for routing decisions."""

    query: str
    params: tuple[Any, ...] = ()
    shard_key: str | None = None       # Column to extract from query
    shard_key_value: Any = None        # Explicit shard key value (overrides extraction)
    transaction_shard: str | None = None  # If inside a transaction, pinned shard ID
    read_only: bool = False


# ---------------------------------------------------------------------------
# SQL shard key extraction (lightweight — not a full parser)
# ---------------------------------------------------------------------------

# Matches: WHERE shard_key = <value> or WHERE shard_key IN (...)
_WHERE_EQ_PATTERN = re.compile(
    r"WHERE\s+.*?\b{key}\b\s*=\s*(?:'([^']*)'|(\d+)|(%s|\$\d+|\?))",
    re.IGNORECASE | re.DOTALL,
)

# DDL patterns that should be broadcast
_DDL_KEYWORDS = {"CREATE", "ALTER", "DROP", "TRUNCATE", "GRANT", "REVOKE"}


def _extract_shard_key(query: str, shard_key: str, params: tuple[Any, ...]) -> Any | None:
    """Attempt to extract the shard key value from a SQL query.

    This is a best-effort heuristic parser. For complex queries, users
    should provide the shard key value explicitly via QueryContext.

    Returns:
        The extracted key value, or None if extraction fails.
    """
    pattern = re.compile(
        rf"\b{re.escape(shard_key)}\b\s*=\s*(?:'([^']*)'|(\d+)|(%s|\$(\d+)|\?))",
        re.IGNORECASE,
    )
    match = pattern.search(query)
    if not match:
        return None

    # String literal: 'value'
    if match.group(1) is not None:
        return match.group(1)

    # Numeric literal: 42
    if match.group(2) is not None:
        return int(match.group(2))

    # Parameterized placeholder
    if match.group(3) is not None:
        placeholder = match.group(3)

        # Positional ($1, $2, ...) — asyncpg style
        if match.group(4) is not None:
            idx = int(match.group(4)) - 1  # $1 → index 0
            if 0 <= idx < len(params):
                return params[idx]
            return None

        # %s or ? — sequential positional
        # Count how many placeholders appear before this one
        prefix = query[: match.start(3)]
        prior_count = prefix.count("%s") + prefix.count("?")
        if prior_count < len(params):
            return params[prior_count]

    return None


def _is_ddl(query: str) -> bool:
    """Check if a query is a DDL statement that should be broadcast."""
    first_word = query.strip().split()[0].upper() if query.strip() else ""
    return first_word in _DDL_KEYWORDS


class Router:
    """
    Query router implementing the MuxDB Layer 1 routing logic.

    The router inspects each query to determine which shard(s) it targets,
    using the configured shard key and the ShardMap's resolution strategy.

    For queries where the shard key cannot be extracted (e.g. complex joins),
    the router falls back to scatter-gather across all shards.

    Example::

        router = Router(config)
        decision = router.route("SELECT * FROM orders WHERE user_id = 42")
        # decision.type == RouteType.SINGLE
        # decision.targets == [ShardInfo(id='shard_2', ...)]
    """

    def __init__(self, config: MuxConfig, shard_map: ShardMap | None = None) -> None:
        self._config = config
        self._shard_key = config.cluster.shard_key
        self._scatter_enabled = config.routing.scatter_gather
        self._shard_map = shard_map or create_shard_map(config)

    @property
    def shard_map(self) -> ShardMap:
        """Access the underlying shard map."""
        return self._shard_map

    def route(self, ctx: QueryContext | str, params: tuple[Any, ...] = ()) -> RouteDecision:
        """Route a query to target shard(s).

        Args:
            ctx: A :class:`QueryContext` or a raw SQL string.
            params: Query parameters (used only when ctx is a string).

        Returns:
            A :class:`RouteDecision` describing where and how to execute.

        Raises:
            RoutingError: If routing fails and scatter-gather is disabled.
            CrossShardTransactionError: If a transaction is pinned and the
                query targets a different shard.
        """
        t0 = time.monotonic()

        if isinstance(ctx, str):
            ctx = QueryContext(query=ctx, params=params)

        decision = self._resolve(ctx)
        elapsed_us = (time.monotonic() - t0) * 1_000_000

        decision = RouteDecision(
            type=decision.type,
            targets=decision.targets,
            shard_key_value=decision.shard_key_value,
            query=ctx.query,
            params=ctx.params,
            elapsed_us=elapsed_us,
        )

        logger.debug(
            "router.route",
            route_type=decision.type.name,
            targets=[t.id for t in decision.targets],
            shard_key_value=decision.shard_key_value,
            elapsed_us=f"{elapsed_us:.1f}",
        )

        return decision

    def _resolve(self, ctx: QueryContext) -> RouteDecision:
        """Core resolution logic."""

        # 1. DDL → broadcast to all shards
        if _is_ddl(ctx.query):
            return RouteDecision(
                type=RouteType.BROADCAST,
                targets=self._shard_map.shards,
            )

        # 2. Transaction pinning — enforce single-shard
        if ctx.transaction_shard:
            shard = self._shard_map.get_shard(ctx.transaction_shard)
            return RouteDecision(
                type=RouteType.SINGLE,
                targets=[shard],
                shard_key_value=ctx.shard_key_value,
            )

        # 3. Explicit shard key value provided
        if ctx.shard_key_value is not None:
            shard = self._shard_map.resolve(ctx.shard_key_value)
            return RouteDecision(
                type=RouteType.SINGLE,
                targets=[shard],
                shard_key_value=ctx.shard_key_value,
            )

        # 4. Extract shard key from query
        effective_key = ctx.shard_key or self._shard_key
        extracted = _extract_shard_key(ctx.query, effective_key, ctx.params)

        if extracted is not None:
            shard = self._shard_map.resolve(extracted)
            return RouteDecision(
                type=RouteType.SINGLE,
                targets=[shard],
                shard_key_value=extracted,
            )

        # 5. Fallback: scatter-gather
        if self._scatter_enabled:
            return RouteDecision(
                type=RouteType.SCATTER,
                targets=self._shard_map.shards,
            )

        raise RoutingError(
            f"Cannot determine shard for query and scatter_gather is disabled. "
            f"Provide a '{effective_key}' value in the WHERE clause or set "
            f"routing.scatter_gather=true in config.",
            details={"query_prefix": ctx.query[:80]},
        )

    def route_key(self, key: str | int) -> ShardInfo:
        """Directly resolve a shard key to its owning shard.

        Convenience method that bypasses query parsing.
        """
        return self._shard_map.resolve(key)

    def route_keys(self, keys: Sequence[str | int]) -> dict[str, list[str | int]]:
        """Group multiple keys by their target shard.

        Returns:
            Dict mapping shard_id → list of keys.
        """
        return self._shard_map.resolve_many(keys)

File: muxdb-js/src/router.ts
/**
 * Query router — Layer 1 core.
 *
 * Parses queries, extracts shard keys, resolves targets via ShardMap,
 * and orchestrates scatter-gather for cross-shard queries.
 */

import { RoutingError, CrossShardTransactionError } from "./errors.js";
import type { MuxConfig } from "./config.js";
import { ShardMap, type ShardInfo } from "./shard-map.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export enum RouteType {
  SINGLE = "SINGLE",
  SCATTER = "SCATTER",
  BROADCAST = "BROADCAST",
}

export interface RouteDecision {
  readonly type: RouteType;
  readonly targets: readonly ShardInfo[];
  readonly shardKeyValue: string | number | null;
  readonly query: string;
  readonly params: unknown[];
  readonly elapsedUs: number;
}

export interface QueryContext {
  readonly query: string;
  readonly params: unknown[];
  readonly shardKey?: string;
  readonly shardKeyValue?: string | number;
  readonly transactionShard?: string;
  readonly readOnly?: boolean;
}

// ---------------------------------------------------------------------------
// SQL shard key extraction
// ---------------------------------------------------------------------------

const DDL_KEYWORDS = new Set(["CREATE", "ALTER", "DROP", "TRUNCATE", "GRANT", "REVOKE"]);

function extractShardKey(
  query: string,
  shardKey: string,
  params: unknown[],
): string | number | null {
  const escapedKey = shardKey.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pattern = new RegExp(
    `\\b${escapedKey}\\b\\s*=\\s*(?:'([^']*)'|(\\d+)|(\\%s|\\$(\\d+)|\\?))`,
    "i",
  );

  const match = pattern.exec(query);
  if (!match) return null;

  // String literal
  if (match[1] !== undefined) return match[1];

  // Numeric literal
  if (match[2] !== undefined) return parseInt(match[2], 10);

  // Parameterized placeholder
  if (match[3] !== undefined) {
    // $N style (asyncpg / pg)
    if (match[4] !== undefined) {
      const idx = parseInt(match[4], 10) - 1;
      if (idx >= 0 && idx < params.length) return params[idx] as string | number;
      return null;
    }

    // %s or ? — count preceding placeholders
    const prefix = query.substring(0, match.index);
    const priorCount = (prefix.match(/%s|\?/g) ?? []).length;
    if (priorCount < params.length) return params[priorCount] as string | number;
  }

  return null;
}

function isDDL(query: string): boolean {
  const firstWord = query.trim().split(/\s+/)[0]?.toUpperCase() ?? "";
  return DDL_KEYWORDS.has(firstWord);
}

// ---------------------------------------------------------------------------
// Router
// ---------------------------------------------------------------------------

export class Router {
  private readonly _shardKey: string;
  private readonly _scatterEnabled: boolean;
  private readonly _shardMap: ShardMap;

  constructor(config: MuxConfig, shardMap?: ShardMap) {
    this._shardKey = config.cluster.shardKey;
    this._scatterEnabled = config.routing.scatterGather;
    this._shardMap = shardMap ?? ShardMap.fromConfig(config);
  }

  get shardMap(): ShardMap {
    return this._shardMap;
  }

  route(ctxOrQuery: QueryContext | string, params: unknown[] = []): RouteDecision {
    const t0 = performance.now();

    const ctx: QueryContext =
      typeof ctxOrQuery === "string"
        ? { query: ctxOrQuery, params }
        : ctxOrQuery;

    const decision = this._resolve(ctx);
    const elapsedUs = (performance.now() - t0) * 1000;

    return {
      ...decision,
      query: ctx.query,
      params: ctx.params,
      elapsedUs,
    };
  }

  routeKey(key: string | number): ShardInfo {
    return this._shardMap.resolve(key);
  }

  routeKeys(keys: (string | number)[]): Map<string, (string | number)[]> {
    return this._shardMap.resolveMany(keys);
  }

  private _resolve(ctx: QueryContext): Omit<RouteDecision, "query" | "params" | "elapsedUs"> {
    // 1. DDL → broadcast
    if (isDDL(ctx.query)) {
      return {
        type: RouteType.BROADCAST,
        targets: this._shardMap.shards,
        shardKeyValue: null,
      };
    }

    // 2. Transaction pinning
    if (ctx.transactionShard) {
      const shard = this._shardMap.getShard(ctx.transactionShard);
      return {
        type: RouteType.SINGLE,
        targets: [shard],
        shardKeyValue: ctx.shardKeyValue ?? null,
      };
    }

    // 3. Explicit shard key value
    if (ctx.shardKeyValue !== undefined) {
      const shard = this._shardMap.resolve(ctx.shardKeyValue);
      return {
        type: RouteType.SINGLE,
        targets: [shard],
        shardKeyValue: ctx.shardKeyValue,
      };
    }

    // 4. Extract from query
    const effectiveKey = ctx.shardKey ?? this._shardKey;
    const extracted = extractShardKey(ctx.query, effectiveKey, ctx.params);

    if (extracted !== null) {
      const shard = this._shardMap.resolve(extracted);
      return {
        type: RouteType.SINGLE,
        targets: [shard],
        shardKeyValue: extracted,
      };
    }

    // 5. Fallback: scatter-gather
    if (this._scatterEnabled) {
      return {
        type: RouteType.SCATTER,
        targets: this._shardMap.shards,
        shardKeyValue: null,
      };
    }

    throw new RoutingError(
      `Cannot determine shard for query and scatter_gather is disabled. ` +
        `Provide a '${effectiveKey}' value in the WHERE clause or set ` +
        `routing.scatterGather=true in config.`,
      { query_prefix: ctx.query.substring(0, 80) },
    );
  }
}
